eventual-cassandra/src/main/scala/com/evolutiongaming/kafka/journal/eventual/cassandra/SnapshotCassandra.scala

```diff
@@ -0,0 +1,182 @@
+    origin: Option[Origin],
+    cassandraClusterOf: CassandraClusterOf[F]
+  ): Resource[F, SnapshotStoreFlat[F]] = {
+
+    def store(implicit cassandraCluster: CassandraCluster[F], cassandraSession: CassandraSession[F]) = {
+      of(config.schema, origin, config.consistencyConfig)
+    }
+
+    for {
+      cassandraCluster <- CassandraCluster.of[F](config.client, cassandraClusterOf, config.retries)
+      cassandraSession <- cassandraCluster.session
+      store <- store(cassandraCluster, cassandraSession).toResource
+    } yield store
+
+  }
+
+  def of[F[_]: Temporal: Parallel: CassandraCluster: CassandraSession: LogOf](
+    schemaConfig: SchemaConfig,
+    origin: Option[Origin],
+    consistencyConfig: ConsistencyConfig
+  ): F[SnapshotStoreFlat[F]] =
+    for {
+      schema <- SetupSchema[F](schemaConfig, origin, consistencyConfig)
+      statements <- Statements.of[F](schema, consistencyConfig)
+    } yield SnapshotCassandra(statements)
+
+  private sealed abstract class Main
+
+  def apply[F[_]: MonadThrow](statements: Statements[F]): SnapshotStoreFlat[F] = {
+    new Main with SnapshotStoreFlat[F] {
+
+      // we do not use segments for now
+      val segmentNr = SegmentNr.min
+
+      def save(key: Key, snapshot: SnapshotRecord[EventualPayloadAndType]): F[Unit] = {
+        statements.selectMetadata(key, segmentNr).flatMap {
+          // such snapshot is already saved, overwrite
+          case s if s.values.exists { case (seqNr, _) => snapshot.snapshot.seqNr == seqNr } =>
+            update(key, segmentNr, s, snapshot)
+          // there is a free place to add a snapshot
+          case s if s.size < BufferSize => insert(key, segmentNr, s, snapshot)
+          // all rows are taken, we have to update one of them
+          case s => replace(key, segmentNr, s, snapshot)
+        }
+      }
+
+      def insert(
+        key: Key,
+        segmentNr: SegmentNr,
+        savedSnapshots: Map[BufferNr, (SeqNr, Instant)],
+        snapshot: SnapshotRecord[EventualPayloadAndType]
+      ): F[Unit] = {
+        val allBufferNrs = BufferNr.listOf(BufferSize)
+        val takenBufferNrs = savedSnapshots.keySet
+        val freeBufferNr = allBufferNrs.find(bufferNr => !takenBufferNrs.contains(bufferNr))
+        MonadThrow[F].fromOption(freeBufferNr, SnapshotStoreError("Could not find a free key")).flatMap { bufferNr =>
+          val wasApplied = statements.insertRecord(key, segmentNr, bufferNr, snapshot)
+          wasApplied.flatMap { wasApplied =>
+            // TODO: consider adding circuit breaker here
+            if (wasApplied) ().pure[F] else save(key, snapshot)
+          }
+        }
+      }
+
+      def replace(
+        key: Key,
+        segmentNr: SegmentNr,
+        savedSnapshots: Map[BufferNr, (SeqNr, Instant)],
+        insertSnapshot: SnapshotRecord[EventualPayloadAndType]
+      ): F[Unit] = {
+        val sortedSnapshots = savedSnapshots.toList.sortBy { case (_, (seqNr, timestamp)) => (seqNr, timestamp) }
```

Comment:
no need to sort collection, just `minBy`

Reply:
is it correct to include `timestamp` here?

Reply:
Technically, it should never happen, i.e. there should not be two snapshots with the same `seqNr`. :thinking: